### PR TITLE
feat: dates filtering implementation [DHIS2-13794][DHIS2-13796][DHIS2-13829]

### PR DESCRIPTION
**Pull request by gnespolino**

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        if ( uidWithOffsets.size() == 1 ) // ie.: fgrg
+        {
+            assertDimensionIdHasNoOffset( uidWithOffsets.get( 0 ) );
+            return DimensionIdentifier.of( emptyElementWithOffset(), emptyElementWithOffset(),
+                uidWithOffsets.get( 0 ).getElement() );
+        }
+
+        throw new IllegalArgumentException( "Malformed parameter: " + fullDimensionId );
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java:72]: I think we could use the standard exception that accepts a code in the format E-7xxx.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            .map( DimensionIdentifierConverterSupport::elementWithOffsetByString )
+            .collect( toList() );
+    }
+
+    private static void assertDimensionIdHasNoOffset(
+        DimensionIdentifier.ElementWithOffset<StringUid> dimensionIdWithOffset )
+    {
+        if ( dimensionIdWithOffset.hasOffset() )
+        {
+            throw new IllegalArgumentException( "Only program and program stage can have offsets" );
```
Comment (maikelarabori) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java:87]: Same here: "I think we could use the standard exception that accepts a code in the format E-7xxx."

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/AbstractCondition.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    protected Renderable getEnrollmentCondition()
+    {
+        // Default implementation, subclasses will override where needed
+        throw new IllegalStateException( "Not supported for enrollment dimension" );
+    }
+
+    private String getEnrollmentSubQuery( String programUid )
+    {
+        return " (SELECT *" +
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/AbstractCondition.java:91]: SQL statements should be all lowercase per our coding standards.
The same applies to all other SQL statements in this PR.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/AbstractCondition.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        // Default implementation, subclasses will override where needed
+        throw new IllegalStateException( "Not supported for event dimension" );
+    }
+
+    private String getEventSubQuery( String programStageUid )
+    {
+        return " (SELECT *" +
+            " from (select *, row_number() over (partition by programinstanceuid order by executiondate desc) as rn" +
+            " from analytics_tei_events_" + queryContext.getTetTableSuffix() +
+            " where programstageuid = " + queryContext.bindParamAndGetIndex( programStageUid ) + " " +
```
Comment (maikelarabori) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/tei/query/AbstractCondition.java:123]: We can remove this extra space concatenation at the end and add it to `" and "`

**Review by maikelarabori — APPROVED** [review]
Nice changes/clean-up @gnespolino. I left some minor comments.

Cheers,
Maikel

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java (hunk 1 above)
Comment (gnespolino) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java:72]: I think we should agree on a strategy to handle exceptions in CPL. I'll leave this as it is, for now, we'll agree on handling all edge cases (ControllerAdvice vs Validator) and refactor this code accordingly.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+@NoArgsConstructor( access = AccessLevel.PRIVATE )
+public class DimensionIdentifierConverterSupport
+{
+
+    private static final Character DIMENSION_SEPARATOR = '.';
+
+    public static DimensionIdentifier<StringUid, StringUid, StringUid> fromFullDimensionId( String fullDimensionId )
+    {
+        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets = parseFullDimensionId( fullDimensionId );
+
```
Comment (d-bernat) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java:49]: It is ok, but what about try this:
```
public static DimensionIdentifier<StringUid,` StringUid, StringUid> fromFullDimensionId( String fullDimensionId )
    {
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets = parseFullDimensionId( fullDimensionId );

        return DimensionIdentifier.of( getProgram(uidWithOffsets), getStage(uidWithOffsets),
                getDimension(uidWithOffsets) );
    }

```
parseFullDimensionId method can throw 
        throw new IllegalArgumentException( "Malformed parameter: " + fullDimensionId );

**Review by d-bernat — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java (hunk 5 above)
Comment (gnespolino) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/dimension/DimensionIdentifierConverterSupport.java:49]: An alternative would look like this:

```java
    public static DimensionIdentifier<StringUid, StringUid, StringUid> fromFullDimensionId( String fullDimensionId )
    {
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets = parseFullDimensionId( fullDimensionId );

        return DimensionIdentifier.of(
            getProgram( uidWithOffsets ),
            getProgramStage( uidWithOffsets ),
            getDimension( uidWithOffsets ) );
    }

    private static DimensionIdentifier.ElementWithOffset<StringUid> getProgram(
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets )
    {
        if ( uidWithOffsets.size() == 2 || uidWithOffsets.size() == 3 )
        {
            return uidWithOffsets.get( 0 );
        }
        else if ( uidWithOffsets.size() == 1 )
        {
            return emptyElementWithOffset();
        }
        throw new IllegalArgumentException( "Invalid dimension identifier: " + uidWithOffsets );
    }

    private static DimensionIdentifier.ElementWithOffset<StringUid> getProgramStage(
        List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets )
    {
        if ( uidWithOffsets.size() == 3 )
        {
            return uidWithOffsets.get( 1 );
        }
        else if ( uidWithOffsets.size() == 2 || uidWithOffsets.size() == 1 )
        {
            return emptyElementWithOffset();
        }
        else
        {
            throw new IllegalArgumentException( "Invalid dimension identifier: " + uidWithOffsets );
        }
    }

    private static StringUid getDimension( List<DimensionIdentifier.ElementWithOffset<StringUid>> uidWithOffsets )
    {
        DimensionIdentifier.ElementWithOffset<StringUid> dimension;
        if ( uidWithOffsets.size() == 3 )
        {
            dimension = uidWithOffsets.get( 2 );
        }
        else if ( uidWithOffsets.size() == 2 )
        {
            dimension = uidWithOffsets.get( 1 );
        }
        else if ( uidWithOffsets.size() == 1 )
        {
            dimension = uidWithOffsets.get( 0 );
        }
        else
        {
            throw new IllegalArgumentException( "Invalid dimension identifier: " + uidWithOffsets );
        }
        assertDimensionIdHasNoOffset( dimension );
        return dimension.getElement();
    }
```

**Pull request merged**